src/driver-asus.c

```diff
@@ -0,0 +1,516 @@
+
+	// get device properties from configuration file
+	profile_count = ratbag_device_data_asus_get_profile_count(device->data);
+	led_count = ratbag_device_data_asus_get_led_count(device->data);
+	mapping = ratbag_device_data_asus_get_button_mapping(device->data);
+
+	// get current profile id
+	profile_id = 0;
+	profile_data = zalloc(sizeof(*profile_data));
+	error = asus_get_profile_data(device, profile_data);
+	if (!error) {
+		profile_id = profile_data->profile_id;
+		log_debug(device->ratbag, "Initial profile is %d\n", profile_id);
+	}
+	free(profile_data);
+	if (error)
+		return error;
+
+	ratbag_device_for_each_profile(device, profile) {
+	if (!profile->dirty)
+		continue;
+
+	log_debug(device->ratbag, "Profile %d changed\n", profile->index);
+
+	// switch profile
+	if (profile->index != profile_id) {
+		log_debug(device->ratbag, "Switching to profile %d\n", profile->index);
+		error = asus_set_profile(device, profile->index);
+		if (error)
+			return error;
+	}
+
+	// set buttons
+	ratbag_profile_for_each_button(profile, button) {
+		if (!button->dirty)
+			continue;
+
+		asus_index = mapping[button->index] != -1 ? mapping[button->index] : ASUS_BUTTON_MAPPING_ORDER[button->index];
+		if (asus_index == -1) {
+			log_debug(device->ratbag, "No mapping for button %d\n", button->index);
+			continue;
+		}
+
+		log_debug(device->ratbag, "Button %d (%d) changed\n", button->index, asus_index);
+		error = 0;
+
+		switch(button->action.type) {
+		case RATBAG_BUTTON_ACTION_TYPE_NONE:
+			error = asus_set_button_action(
+				device, asus_index,
+				ASUS_BUTTON_CODE_DISABLED,
+				ASUS_BUTTON_ACTION_TYPE_BUTTON);
+			break;
+
+		case RATBAG_BUTTON_ACTION_TYPE_KEY:
+			// linux code to asus code
+			asus_index_to = asus_find_key_index_by_code(button->action.action.key.key);
+			if (asus_index_to >= 0)
+				error = asus_set_button_action(
+					device, asus_index,
+					asus_index_to,
+					ASUS_BUTTON_ACTION_TYPE_KEY);
+		break;
+
+		case RATBAG_BUTTON_ACTION_TYPE_BUTTON:
+		case RATBAG_BUTTON_ACTION_TYPE_SPECIAL:
+			asus_index_to = asus_find_button_index_by_action(button->action);
+			if (asus_index_to >= 0)
+				error = asus_set_button_action(
+					device, asus_index,
+					ASUS_BUTTON_MAPPING[asus_index_to].asus_code,
+					ASUS_BUTTON_ACTION_TYPE_BUTTON);
+		break;
+
+		default:
+			break;
+		}
+
+		if (error)
+			return error;
+	}
+
+	// set extra settings
+	if (profile->rate_dirty) {
+		log_debug(device->ratbag, "Polling rate changed to %d Hz\n", profile->hz);
+		error = asus_set_polling_rate(device, profile->hz);
+		if (error)
+			return error;
+	}
+
+	// set DPIs
+	ratbag_profile_for_each_resolution(profile, resolution) {
+		if (!resolution->dirty)
+			continue;
+
+		log_debug(
+			device->ratbag, "Resolution %d changed to %d\n",
+			resolution->index, resolution->dpi_x);
+
+		error = asus_set_dpi(device, resolution->index, resolution->dpi_x);
+		if (error)
+			return error;
+	}
+
+	// set LEDs
+	if (led_count) {
```

Comment:
is this check necessary? `ratbag_profile_for_each_led` should just be an empty loop if we don't have LEDs.